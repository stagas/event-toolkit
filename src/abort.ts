import type { Fn } from 'everyday-types'
import { bool, toFluent } from 'to-fluent'

export class AbortOptions {
  throw = bool
  latest = bool
  timeout?: number
}

export const abort = toFluent(
  AbortOptions,
  options =>
    <P extends any[], R extends Promise<any>>(fn: (signal: AbortSignal) => Fn<P, R>) => {
      let ctrl: AbortController
      return async function(this: any, ...args: P) {
        if (options.latest && ctrl) ctrl.abort()
        ctrl = new AbortController()
        if (options.timeout != null) {
          const timeoutError = new Error('Timed out')
          try {
            return await Promise.race([
              new Promise((_, reject) => setTimeout(reject, options.timeout, timeoutError)),
              fn(ctrl.signal).apply(this, args),
            ])
          } catch (error) {
            if (error === timeoutError) {
              ctrl.abort()
              if (options.throw)
                throw error
              else
                console.warn(error)
            } else {
              if (options.throw)
                throw error
              else
                console.warn(error)
            }
          }
        } else {
          return fn(ctrl.signal).apply(this, args)
        }
      }
    }
)
