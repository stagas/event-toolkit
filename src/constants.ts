export const MouseButton = {
  Left: 1,
  Right: 2,
  Middle: 4,
}
