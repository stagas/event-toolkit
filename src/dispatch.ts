import type { DetailOf, EventKeys, Prefix, StringOf } from 'everyday-types'
import { bool, Fluent, toFluent } from 'to-fluent'

export class DispatchOptions implements CustomEventInit {
  bubbles = bool
  cancelable = bool
  composed = bool
}

export type Dispatch<T> = T & Fluent<T, Required<DispatchOptions>>

export const dispatch = toFluent(
  DispatchOptions,
  options => (<T extends EventTarget, K extends EventKeys<T>>(
    el: T,
    nameOrEvent: StringOf<K> | Event,
    detail?: DetailOf<T, Prefix<'on', K>>,
    init?: CustomEventInit,
  ) =>
    el.dispatchEvent(
      nameOrEvent instanceof Event
        ? nameOrEvent
        : new CustomEvent(nameOrEvent, { detail, ...init, ...options })
    ))
)
