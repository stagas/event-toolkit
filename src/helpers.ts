export const chain = (first: () => Promise<any>, ...rest: (() => Promise<any>)[]) =>
  () => rest.reduce((p, n) => p.then(n), first())
