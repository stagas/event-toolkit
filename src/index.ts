export * from './abort'
export * from './constants'
export * from './dispatch'
export * from './event'
export * from './helpers'
export * from './on'
export * from './queue'
export * from './task'
