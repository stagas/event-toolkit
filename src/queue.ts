import type { Fn } from 'everyday-types'
import { shallowEqual } from 'everyday-utils'
import { bool, toFluent } from 'to-fluent'
import { taskGroup, taskRun, Task } from './task'

export interface Hooks {
  before?: () => void
  after?: () => void
}

export class QueueOptions {
  first = bool
  last = bool
  next = bool

  raf = bool
  task = bool
  time = bool

  atomic = bool
  concurrency?: number

  debounce?: number
  throttle?: number

  hooks?: Hooks
}

export const wrapQueue = (options: Partial<QueueOptions> = {}) =>
  <P extends any[], R>(queueFn: Fn<P, R>): Fn<P, R extends Promise<any> ? R : Promise<R>> => {
    const initialOptions = { ...options }
    const queued: Task[] = []
    let queueOuterFn: Fn<[Fn<any, any>], any>
    let id: any
    let last: any
    let runs = false

    if (options.hooks) {
      if (options.hooks.before || options.hooks.after) {
        const real = queueFn
        queueFn = function queueFn(this: any, ...args: P) {
          options.hooks!.before?.()
          const result = real.apply(this, args)
          options.hooks!.after?.()
          return result
        }
      }
    }

    if (options.raf) queueOuterFn = requestAnimationFrame
    else if (options.task) queueOuterFn = queueMicrotask
    else if (options.time) queueOuterFn = setTimeout
    else if (options.throttle != null) {
      queueOuterFn = fn => {
        runs = true
        setTimeout(() => {
          if (!queued.length) runs = false
          fn()
        }, options.throttle)
      }
    } else if (options.debounce != null) {
      queueOuterFn = fn => {
        clearTimeout(id)
        id = setTimeout(fn, options.debounce)
      }
    } else if (options.atomic)
      queueOuterFn = fn => fn()
    else {
      // No queue function provided, return identity.
      // This is used when extending this in `event`.
      Object.assign(queueFn, {
        fn: queueFn,
        update: () => queueFn
      })
      return queueFn as any
    }

    if (options.first == null && options.last == null) {
      if (options.throttle != null) {
        options.first = true
        options.last = true
      }
      else if (options.debounce != null)
        options.last = true
      else
        options.next = true
    }

    let runningTasks = 0

    function queueNext() {
      let task: Task

      if (queued.length) {
        if (options.atomic) {
          task = queued.shift()!
          taskRun(task)
            .catch((_error: Error) => {
              //!warn _error
            })
            .finally(() => {
              if (options.concurrency) {
                runningTasks--
              }
              queueOuterFn(queueNext)
            })
          if (options.concurrency) {
            runningTasks++
            if (runningTasks < options.concurrency) {
              queueOuterFn(queueNext)
            }
          }
          return
        }
        if (options.last) {
          if (options.next) {
            const left = queued.splice(0, queued.length - 1)
            task = left.pop() ?? queued.pop()!
            taskGroup(task, left)
            last = taskRun(task)
            if (queued.length || options.throttle) {
              queueOuterFn(queueNext)
            }
            return
          } else {
            task = queued.pop()!
            taskGroup(task, queued.splice(0))
            last = taskRun(task)
            if (options.throttle) {
              queueOuterFn(queueNext)
              return
            }
          }
        } else if (options.next) {
          task = queued.shift()!
          taskGroup(task, queued.splice(0, queued.length - 1))
          queueOuterFn(queueNext)
          last = taskRun(task)
          return
        } else {
          task = Task()
          taskGroup(task, queued.splice(0))
          task.resolve(last)
        }
      }
      runs = false
    }

    function queueWrap(this: any, ...args: P) {
      //!? 'wrap called'
      const task = Task(queueFn, this, args)

      if (!runs && options.first) {
        runs = true
        if (!queued.length) {
          last = taskRun(task)
          queueOuterFn(queueNext)
          return task.promise
        }
      }

      queued.push(task)

      if (!runs || options.debounce) {
        runs = true
        queueOuterFn(queueNext)
      }

      return task.promise
    }

    queueWrap.fn = queueFn
    queueWrap.options = initialOptions

    queueWrap.update = (newFn: Fn<P, R>, newOptions: QueueOptions) => {
      //!? 'updating fn'
      if (!shallowEqual(initialOptions, newOptions)) {
        //!? 'new options', initialOptions, newOptions
        return newFn
      }
      queueFn = newFn
      //!? 'updated and returned previous wrapped'
      return queueWrap
    }

    return queueWrap as any
  }

/**
 * Decorate function `fn` with a queue function.
 *
 * The decorated function will returns a `Promise` that resolves with its result.
 * All calls will be resolved with latest result at any given quantum.
 *
 * Options:
 * - `first` => Run only first, then debounce.
 * - `last` => Run last (default behavior when using `debounce`).
 * - `next` => Run final subsequent call on next quantum (default behavior when nothing is set and not a `debounce`).
 *
 * - `raf` => Queue with `requestAnimationFrame`.
 * - `task` => Queue with `queueMicrotask`.
 * - `time` => Queue with `setTimeout`.
 *
 * - `debounce(ms)` => Debounce at specified `ms`.
 * - `throttle(ms)` => Throttle at specified `ms`.
 *
 * ```ts
 * fn = (x: number) => console.log(x)
 *
 * const cb = queue.task((x: number) => count += x)
 * cb(1) // passes (after task)
 * cb(2) // discarded
 * cb(3) // discarded
 * cb(4) // passes
 *
 * const cb = queue.task.last((x: number) => count += x)
 * cb(1) // discarded
 * cb(2) // discarded
 * cb(3) // discarded
 * cb(4) // passes
 *
 * const cb = queue.task.first((x: number) => count += x)
 * cb(1) // passes (before task)
 * cb(2) // discarded
 * cb(3) // discarded
 * cb(4) // discarded
 *
 * const cb = queue.task.first.last((x: number) => count += x)
 * cb(1) // passes (before task)
 * cb(2) // discarded
 * cb(3) // discarded
 * cb(4) // passes
 *
 * const cb = queue.task.first.last.next((x: number) => count += x)
 * cb(1) // passes (before task)
 * cb(2) // discarded
 * cb(3) // passes (after task)
 * cb(4) // passes (next task)
 *
 * const cb = queue.task.last.next((x: number) => count += x)
 * cb(1) // discarded
 * cb(2) // discarded
 * cb(3) // passes (after task)
 * cb(4) // passes (next task)
 * ```
 */
export const queue = toFluent(QueueOptions, wrapQueue)
