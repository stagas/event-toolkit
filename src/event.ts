import { EventHandler, Target } from 'everyday-types'
import { bool, toFluent } from 'to-fluent'
import { QueueOptions, wrapQueue } from './queue'

export class EventOptions extends QueueOptions {
  prevent = bool
  stop = bool
  immediate = bool
}

export const wrapEvent = (options: EventOptions) =>
  <T extends Target, E extends Event>(fn: EventHandler<T, E> = () => {}) =>
    wrapQueue(options)(
      options.prevent || options.stop || options.immediate
        ? function(this: T, e: E) {
          if (options.prevent) e.preventDefault()
          if (options.stop) {
            options.immediate
              ? e.stopImmediatePropagation()
              : e.stopPropagation()
          }
          return fn.call(this, e as any)
        }
        : fn
    )

/**
 * Decorates event handler `fn`.
 *
 * Options:
 * - `prevent` => `event.preventDefault()`
 * - `stop` => `event.stopPropagation()`
 * - `stop.immediate` => `event.stopImmediatePropagation()`
 *
 * Options inherited from `queue`:
 * - `first` => Run only first, then debounce.
 * - `last` => Run last (default behavior when using `debounce`).
 * - `next` => Run final subsequent call on next quantum (default behavior when nothing is set and not a `debounce`).
 *
 * - `raf` => Queue with `requestAnimationFrame`.
 * - `task` => Queue with `queueMicrotask`.
 * - `time` => Queue with `setTimeout`.
 *
 * - `debounce(ms)` => Debounce at specified `ms`.
 * - `throttle(ms)` => Throttle at specified `ms`.
 *
 * ```ts
 * btn.onclick = event(fn)
 * btn.onclick = event.prevent(fn)
 * btn.onclick = event.prevent.stop(fn)
 * btn.onclick = event.stop.immediate(fn)
 *
 * // examples with queue
 * btn.onclick = event.stop.raf(fn)
 * btn.onclick = event.prevent.throttle(50)(fn)
 * btn.onclick = event.debounce(100)(fn)
 * ```
 */
export const event = toFluent(EventOptions, wrapEvent)
