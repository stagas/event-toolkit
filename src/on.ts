import type { EventHandler, EventKeys, EventsOf, Fn, StringOf } from 'everyday-types'
import { Getter } from 'proxy-toolkit'
import { bool, Fluent, toFluent } from 'to-fluent'
import { EventOptions, wrapEvent } from './event'

/**
 * Removes the event listener and returns a promise used for chaining.
 * @public
 */
export type Off = () => Promise<void>
export type On<T> = T & Fluent<T, Required<OnOptions>>
export type OnEvent<T, K extends EventKeys<T>> = On<Fn<[EventHandler<T, EventsOf<T>[K]>?], Off>>
export type OnGetter<T> = { [K in EventKeys<T>]: OnEvent<T, K> }

export class OnOptions extends EventOptions implements AddEventListenerOptions {
  once = bool
  passive = bool
  capture = bool
}

const onEvent = (el: EventTarget, type: string, listener: EventHandler<any, any>, options: AddEventListenerOptions) => {
  el.addEventListener(type, listener, options)
  return () => Promise.resolve(el.removeEventListener(type, listener, options))
}

const onEventFluent = (el: EventTarget, key: string) =>
  toFluent(OnOptions, options => (listener = () => {}) => onEvent(el, key, wrapEvent(options)(listener), options))

/**
 * Adds an event listener for `el` using fluent options.
 *
 * Options:
 * - `capture` => If specified, indicates that events of this type will be dispatched to the registered listener before being dispatched to any `EventTarget` beneath it in the DOM tree.
 * - `once` => If specified, indicates that the listener should be invoked at most once after being added. The listener would be automatically removed when invoked.
 * - `passive` => Indicates that the function specified by listener will never call `preventDefault()`. If a passive listener does call `preventDefault()`, the user agent will do nothing other than generate a console warning. If not specified, defaults to `not.passive` â€“ except that in browsers other than Safari and Internet Explorer, defaults to `passive` for the `wheel`, `mousewheel`, `touchstart` and `touchmove` events.
 * - `signal(ctrl.signal)` => An `AbortSignal`. The listener will be removed when the given `AbortSignal` object's `abort()` method is called. If not specified, no `AbortSignal` is associated with the listener.
 *
 * Options inherited from `event`:
 * - `prevent` => `event.preventDefault()`
 * - `stop` => `event.stopPropagation()`
 * - `stop.immediate` => `event.stopImmediatePropagation()`
 *
 * Options inherited from `queue`:
 * - `first` => Run only first, then debounce.
 * - `last` => Run last (default behavior when using `debounce`).
 * - `next` => Run final subsequent call on next quantum (default behavior when nothing is set and not a `debounce`).
 *
 * - `raf` => Queue with `requestAnimationFrame`.
 * - `task` => Queue with `queueMicrotask`.
 * - `time` => Queue with `setTimeout`.
 *
 * - `debounce(ms)` => Debounce at specified `ms`.
 * - `throttle(ms)` => Throttle at specified `ms`.
 *
 * ```ts
 * on(btn).click(e => console.log('clicked'))
 * on(btn).click.once(e => console.log('clicked once'))
 * on(btn).wheel.not.passive(e => console.log('wheel not passive'))
 *
 * const off = on(btn).pointerdown.capture(e => console.log('pointer down'))
 * off() // removes the listener
 *
 * const offPointerMove = on(window).pointermove(e => console.log('pointer move'))
 * const offPointerUp = on(window).pointerup(e => console.log('pointer up'))
 * offPointerUp(offPointerMove) // remove both listeners shortcut syntax
 *
 * const ctrl = new AbortController()
 * on(window).pointermove.signal(ctrl.signal)(e =>
 *   console.log('runs until aborted'))
 * ctrl.abort() // removes the listener by signaling abort
 *
 * on(btn).click.prevent(e => console.log('prevented default'))
 * ```
 *
 * @public
 */

export function on<T extends EventTarget>(el: T): OnGetter<T>
export function on<T extends EventTarget, K extends EventKeys<T>>(el: T, key: K): OnEvent<T, K>
export function on<T extends EventTarget, K extends EventKeys<T>>(el: T, key?: StringOf<K>) {
  return key != null
    ? onEventFluent(el, key)
    : Getter(key => onEventFluent(el, key))
}
