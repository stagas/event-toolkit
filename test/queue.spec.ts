// @env browser
// @env jsdom
import { abort } from '../src/abort'
import { queue } from '../src/queue'

const wait = (ms: number) => new Promise(resolve => setTimeout(resolve, ms))
let rx = 0
const rnd = () => Math.sin(++rx * 1000) / 2 + 0.5

describe('queue(fn)', () => {
  describe('task', () => {
    it('one', async () => {
      let count = 0
      const cb = queue.task((x: number) => count += x)
      cb(1)
      expect(count).toEqual(0)
      await Promise.resolve()
      expect(count).toEqual(1)
    })

    it('multiple', async () => {
      let count = 0
      const cb = queue.task((x: number) => count += x)
      cb(1) // passes (after task)
      cb(2) // discarded
      cb(3) // discarded
      cb(4) // passes
      expect(count).toEqual(0)
      await Promise.resolve()
      expect(count).toEqual(1)
      await Promise.resolve()
      expect(count).toEqual(5)
    })

    it('last', async () => {
      let count = 0
      const cb = queue.task.last((x: number) => count += x)
      cb(1) // discarded
      cb(2) // discarded
      cb(3) // discarded
      cb(4) // passes
      expect(count).toEqual(0)
      await Promise.resolve()
      expect(count).toEqual(4)
    })

    it('first', async () => {
      let count = 0
      const cb = queue.task.first((x: number) => count += x)
      cb(1) // passes (before task)
      cb(2) // discarded
      cb(3) // discarded
      cb(4) // discarded
      expect(count).toEqual(1)
      await Promise.resolve()
      expect(count).toEqual(1)
      await Promise.resolve()
      expect(count).toEqual(1)
    })

    it('first.last', async () => {
      let count = 0
      const cb = queue.task.first.last((x: number) => count += x)
      cb(1) // passes (before task)
      cb(2) // discarded
      cb(3) // discarded
      cb(4) // passes
      expect(count).toEqual(1)
      await Promise.resolve()
      expect(count).toEqual(5)
      await Promise.resolve()
      expect(count).toEqual(5)
    })

    it('first.last', async () => {
      let count = 0
      const cb = queue.task.first.last.next((x: number) => count += x)
      cb(1) // passes (before task)
      cb(2) // discarded
      cb(3) // passes (after task)
      cb(4) // passes (next task)
      expect(count).toEqual(1)
      await Promise.resolve()
      expect(count).toEqual(4)
      await Promise.resolve()
      expect(count).toEqual(8)
    })

    it('last.next', async () => {
      let count = 0
      const cb = queue.task.last.next((x: number) => count += x)
      cb(1) // discarded
      cb(2) // discarded
      cb(3) // passes (after task)
      cb(4) // passes (next task)
      expect(count).toEqual(0)
      await Promise.resolve()
      expect(count).toEqual(3)
      await Promise.resolve()
      expect(count).toEqual(7)
    })
  })

  describe('async/promise functions', () => {
    it('works', async () => {
      const cb = queue.task(async () => {
        return 'ok'
      })
      expect(await cb()).toEqual('ok')
    })
  })

  describe('raf', () => {
    it('one', async () => {
      let count = 0
      const cb = queue.raf(() => count++)
      cb()
      expect(count).toEqual(0)
      await new Promise(resolve => requestAnimationFrame(resolve))
      expect(count).toEqual(1)
    })

    it('two', async () => {
      let count = 0
      const cb = queue.raf(() => count++)
      cb()
      cb()
      expect(count).toEqual(0)
      await new Promise(resolve => requestAnimationFrame(resolve))
      expect(count).toEqual(1)
      await new Promise(resolve => requestAnimationFrame(resolve))
      expect(count).toEqual(2)
    })

    it('multiple plain', async () => {
      let count = 0
      const cb = queue.raf(() => count++)
      cb()
      cb()
      cb()
      cb()
      expect(count).toEqual(0)
      const results: any = []
      await new Promise(resolve => {
        requestAnimationFrame(() => {
          results.push(count)
          requestAnimationFrame(resolve)
        })
      })
      results.push(count)
      expect(count).toEqual(2)
      expect(results).toMatchSnapshot()
    })

    it('multiple with args', async () => {
      let value: any
      let count = 0
      const cb = queue.raf((x: number) => {
        count++
        value = x
        return x
      })
      cb(1)
      cb(2)
      cb(3)
      cb(4)
      expect(value).toBeUndefined()
      const results: any = []
      await new Promise(resolve => {
        requestAnimationFrame(() => {
          results.push(value)
          requestAnimationFrame(resolve)
        })
      })
      results.push(value)
      expect(count).toBe(2)
      expect(value).toBe(4)
      expect(results).toMatchSnapshot()
    })

    it('resolves with result', async () => {
      let value: any
      let count = 0
      const cb = queue.raf((x: number) => {
        value = x
        count++
        return x
      })
      const p1 = cb(1)
      const p2 = cb(2)
      const p3 = cb(3)
      const result = await cb(4)
      expect(value).toBe(4)
      expect(result).toBe(4)
      expect(await p1).toBe(1)
      expect(await p2).toBe(1)
      expect(await p3).toBe(1)
      expect(count).toBe(2)
    })
  })

  describe('time', () => {
    it('time', async () => {
      let count = 0
      const cb = queue.time(() => count++)
      cb()
      expect(count).toEqual(0)
      await new Promise(resolve => setTimeout(resolve))
      expect(count).toEqual(1)
    })
  })

  describe('throttle', () => {
    it('first === thottle', async () => {
      let count = 0
      const cb = queue.throttle(10)((x: number) => count += x)
      cb(2) // passes
      cb(3) // discarded
      cb(4) // discarded
      expect(count).toEqual(2)
      await wait(11)
      expect(count).toEqual(2)
    })

    it('first.last', async () => {
      let count = 0
      const cb = queue.throttle(10).first.last((x: number) => count += x)
      cb(2) // passes
      cb(3) // discarded
      cb(4) // passes
      expect(count).toEqual(2)
      await wait(11)
      expect(count).toEqual(6)
    })
  })

  describe('debounce', () => {
    it('debounce() = last', async () => {
      let count = 0
      const p = []
      const cb = queue.debounce(5)((x: number) => count += x)
      p.push(cb(2)) // discarded
      await wait(2)
      p.push(cb(3)) // discarded
      await wait(2)
      p.push(cb(4)) // discarded
      await wait(2)
      p.push(cb(5)) // discarded
      await wait(2)
      p.push(cb(6)) // passes(last)
      expect(count).toBe(0)
      await wait(6)
      expect(count).toBe(6)
      await wait(6)
      expect(count).toBe(6)
      expect(await Promise.all(p)).toMatchSnapshot()
      expect(count).toBe(6)
    })

    it('first()', async () => {
      let count = 0
      const p = []
      const cb = queue.debounce(5).first((x: number) => (count += x))
      p.push(cb(2)) // passes (first)
      await wait(2)
      p.push(cb(3)) // discarded
      await wait(2)
      p.push(cb(4)) // discarded
      await wait(2)
      p.push(cb(5)) // discarded
      expect(count).toBe(2)
      await wait(6)
      expect(await Promise.all(p)).toMatchSnapshot()
      expect(count).toBe(2)
      cb(2)
      expect(count).toBe(4)
    })

    it('first.last()', async () => {
      let count = 0
      const p = []
      const cb = queue.debounce(5).first.last((x: number) => (count += x))
      p.push(cb(2)) // passes (first)
      await wait(2)
      p.push(cb(3)) // discarded
      await wait(2)
      p.push(cb(4)) // discarded
      await wait(2)
      p.push(cb(5)) // discarded
      await wait(2)
      p.push(cb(6)) // passes (last)
      expect(count).toBe(2)
      await wait(6)
      expect(count).toBe(8)
      await wait(6)
      expect(await Promise.all(p)).toMatchSnapshot()
      expect(count).toBe(8)
    })

    it('first.last.next()', async () => {
      let count = 0
      const p = []
      const cb = queue.debounce(5).first.last.next((x: number) => (count += x))
      p.push(cb(2)) // passes (first)
      await wait(2)
      p.push(cb(3)) // discarded
      await wait(2)
      p.push(cb(4)) // discarded
      await wait(2)
      p.push(cb(5)) // passes
      await wait(2)
      p.push(cb(6)) // passes (next quantum)
      expect(count).toBe(2)
      await wait(6)
      expect(count).toBe(7)
      await wait(6)
      expect(count).toBe(13)
      expect(await Promise.all(p)).toMatchSnapshot()
      expect(count).toBe(13)
    })

    it('last.next() simple', async () => {
      let count = 0
      const p = []
      const cb = queue.debounce(5).last.next((x: number) => (count += x))
      p.push(cb(2)) // discarded
      await wait(2)
      p.push(cb(3)) // discarded
      await wait(2)
      p.push(cb(4)) // discarded
      await wait(2)
      p.push(cb(5)) // passes
      await wait(2)
      p.push(cb(6)) // passes (next quantum)
      expect(count).toBe(0)
      await wait(6)
      expect(count).toBe(5)
      await wait(6)
      expect(count).toBe(11)
      expect(await Promise.all(p)).toMatchSnapshot()
      expect(count).toBe(11)
    })

    it('last.next() with additional', async () => {
      let count = 0
      const p = []
      const cb = queue.debounce(5).last.next((x: number) => (count += x))
      p.push(cb(2)) // discarded
      await wait(2)
      p.push(cb(3)) // discarded
      await wait(2)
      p.push(cb(4)) // discarded
      await wait(2)
      p.push(cb(5)) // passes
      await wait(2)
      p.push(cb(6)) // passes but is discarded by next
      expect(count).toBe(0)
      await wait(5)
      expect(count).toBe(5)
      p.push(cb(7)) // discards 6
      await wait(2)
      p.push(cb(8)) // passes (last)
      await wait(2)
      p.push(cb(9)) // passes (next quantum)
      await wait(5)
      expect(count).toBe(13)
      await wait(5)
      expect(count).toBe(22)
      expect(await Promise.all(p)).toMatchSnapshot()
      expect(count).toBe(22)
    })
  })

  describe('atomic(fn)', () => {
    it('run in order', async () => {
      const results: any = []
      let i = 0
      const fn = queue.atomic(async () => {
        const res = ++i
        await wait(rnd() * 10)
        results.push(res)
      })
      await Promise.all([
        fn(),
        fn(),
        fn(),
        fn(),
        fn(),
      ])
      expect(results).toMatchSnapshot()
    })

    it('makes the function atomic', async () => {
      const results: any = []
      let i = 0
      const fn = queue.atomic(async () => {
        const a = i
        await new Promise(resolve => setTimeout(resolve, rnd() * 100))
        results.push(a + 1)
        i = a + 1
      })
      fn()
      fn()
      fn()
      await fn()
      expect(i).toEqual(4)
      expect(results).toMatchSnapshot()
    })

    it('receives parameters', async () => {
      const results: any = []
      let i = 0
      const fn = queue.atomic(async (x: number) => {
        const a = i
        await new Promise(resolve => setTimeout(resolve, rnd() * 100))
        results.push(a + 1 + x)
        i = a + 1
      })
      fn(1)
      fn(1)
      fn(1)
      await fn(1)
      expect(i).toEqual(4)
      expect(results).toMatchSnapshot()
    })

    it('alternate', async () => {
      const results: any = []
      let i = 0
      const fn = queue.atomic(async () => {
        const a = i
        await new Promise(resolve => setTimeout(resolve, rnd() * 100))
        results.push(a + 1)
        i = a + 1
      })
      const promises = [fn(), fn(), fn(), fn()]
      await Promise.all(promises)
      expect(i).toEqual(4)
      expect(results).toMatchSnapshot()
    })

    it('returns result', async () => {
      const results: any = []
      let i = 0
      const fn = queue.atomic(async () => {
        const a = i
        await new Promise(resolve => setTimeout(resolve, rnd() * 100))
        results.push(a + 1)
        i = a + 1
        return i
      })
      fn()
      fn()
      fn()
      const result = await fn()
      expect(i).toEqual(4)
      expect(result).toEqual(4)
      expect(results).toMatchSnapshot()
    })

    it('throwing doesnt destroy queue', async () => {
      const results: any = []
      let i = 0
      let times = 0
      const fn = queue.atomic(async () => {
        const a = i
        times++
        if (a === 2) throw new Error('problem')
        await new Promise(resolve => setTimeout(resolve, rnd() * 100))
        results.push(a + 1)
        i = a + 1
        return i
      })
      await Promise.allSettled([fn(), fn(), fn(), fn()])
      expect(i).toEqual(2)
      expect(times).toEqual(4)
      expect(results).toMatchSnapshot()
    })

    it('uses this', async () => {
      const results: any = []
      const obj = {
        i: 0,
        fn: queue.atomic(
          async function(this: any) {
            const a = this.i
            await new Promise(resolve => setTimeout(resolve, rnd() * 100))
            results.push(a + 1)
            this.i = a + 1
            return a
          }
        ),
      }
      await Promise.allSettled([obj.fn(), obj.fn(), obj.fn(), obj.fn()])
      expect(obj.i).toEqual(4)
      expect(results).toMatchSnapshot()
    })

    it('with timeout', async () => {
      const results: any = []
      let i = 0
      let times = 0
      const fn = queue.atomic(
        abort.timeout(200)(signal =>
          async () => {
            const a = i
            times++
            await new Promise(resolve => setTimeout(resolve, a === 2 ? 300 : 100))
            if (signal.aborted) return
            results.push(a + 1)
            i = a + 1
            return i
          }
        )
      )
      const output = await Promise.allSettled([fn(), fn(), fn(), fn()])
      await new Promise(resolve => setTimeout(resolve, 200))
      expect(i).toEqual(2)
      expect(times).toEqual(4)
      expect(results).toMatchSnapshot()
      expect(output).toMatchSnapshot()
    })
  })
})
