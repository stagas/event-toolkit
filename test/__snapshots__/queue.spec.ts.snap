// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queue(fn) atomic(fn) alternate 1`] = `
Array [
  1,
  2,
  3,
  4,
]
`;

exports[`queue(fn) atomic(fn) makes the function atomic 1`] = `
Array [
  1,
  2,
  3,
  4,
]
`;

exports[`queue(fn) atomic(fn) receives parameters 1`] = `
Array [
  2,
  3,
  4,
  5,
]
`;

exports[`queue(fn) atomic(fn) returns result 1`] = `
Array [
  1,
  2,
  3,
  4,
]
`;

exports[`queue(fn) atomic(fn) run in order 1`] = `
Array [
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`queue(fn) atomic(fn) throwing doesnt destroy queue 1`] = `
Array [
  1,
  2,
]
`;

exports[`queue(fn) atomic(fn) uses this 1`] = `
Array [
  1,
  2,
  3,
  4,
]
`;

exports[`queue(fn) atomic(fn) with timeout 1`] = `
Array [
  1,
  2,
]
`;

exports[`queue(fn) atomic(fn) with timeout 2`] = `
Array [
  Object {
    "status": "fulfilled",
    "value": 1,
  },
  Object {
    "status": "fulfilled",
    "value": 2,
  },
  Object {
    "status": "fulfilled",
    "value": undefined,
  },
  Object {
    "status": "fulfilled",
    "value": undefined,
  },
]
`;

exports[`queue(fn) debounce debounce() = last 1`] = `
Array [
  6,
  6,
  6,
  6,
  6,
]
`;

exports[`queue(fn) debounce first() 1`] = `
Array [
  2,
  2,
  2,
  2,
]
`;

exports[`queue(fn) debounce first.last() 1`] = `
Array [
  2,
  8,
  8,
  8,
  8,
]
`;

exports[`queue(fn) debounce first.last.next() 1`] = `
Array [
  2,
  7,
  7,
  7,
  13,
]
`;

exports[`queue(fn) debounce last.next() simple 1`] = `
Array [
  5,
  5,
  5,
  5,
  11,
]
`;

exports[`queue(fn) debounce last.next() with additional 1`] = `
Array [
  5,
  5,
  5,
  5,
  13,
  13,
  13,
  22,
]
`;

exports[`queue(fn) raf multiple plain 1`] = `
Array [
  1,
  2,
]
`;

exports[`queue(fn) raf multiple with args 1`] = `
Array [
  1,
  4,
]
`;
